### DISCUSSION TITLE
Bundler: Bundle List With Selective Deleter?

### DISCUSSION THREAD

**Discussion by TiddlyTweeter** [wikilabs/plugins#61] [Ideas]
@pmario I constantly import JSON list bundles to wiki. What I lack is any methodology to LATER select items from a bundle I imported to DELETE. Here is an example list bundle for _$:/tt/bun/search-en-dictionaries_ ...

> [[--- !This tiddler ---]]
> $:/tt/bun/search-en-dictionaries
> 
> [[--- Search Online English Dictionaries ---]]
> $:/tt/eng/en-dict-search.dat
> $:/tt/eng/en-dict-search.mac
> $:/tt/eng/en-dict-search.hlp
> $:/tt/eng/en-dict-search.tpl

It is great to add that to wikis for folk learning English. BUT, later I want to remove SOME items.

My QUESTION: how could I develop a tool that would ...

1. ... parse the bundle
2. ... ignore comment lines
3. ... for all tiddlers present give a list with a checkbox to (selectively) delete them?

My ideal would be to INCLUDE a deleter in the bundle so that it would read like this ...

> [[--- !This tiddler ---]]
> $:/tt/bun/search-en-dictionaries
> [[--- Selective Deleter For This Bundle ---]]
> **$:/tt/bun/search-en-dictionaries-REMOVER**
> 
> [[--- Search Online English Dictionaries ---]]
> ... etc

You get the idea? It is just a thought. But I think a valid one.
Maybe there is a not too difficult solution?

Best wishes, TT

**Comment by pmario** [comment]
>My ideal would be to INCLUDE a deleter in the bundle so that it would read like this ...
>
>    [[--- !This tiddler ---]]
    $:/tt/bun/search-en-dictionaries
    [[--- Selective Deleter For This Bundle ---]]
    $:/tt/bun/search-en-dictionaries-REMOVER
>
>    [[--- Search Online English Dictionaries ---]]
    ... etc
>
>You get the idea? 

No. ... You don't tell me, what exactly should be "removed". 

>It is just a thought. But I think a valid one.

I think it is a valid thought, but ATM the TW import mechanism only has the possibility to "disable" import and not "delete" existing content.

**Reply by TiddlyTweeter** [reply]
> @pmario ... the TW import mechanism only has the possibility to "disable" import and not "delete" existing content.

Sorry. That is my bad by not making clear what the function might be.

**Reply by TiddlyTweeter** [reply]
@TiddlyTweeter 
> > [[--- !This tiddler ---]]
> > $:/tt/bun/search-en-dictionaries
> > [[--- Selective Deleter For This Bundle ---]]
> > $:/tt/bun/search-en-dictionaries-REMOVER
> > [[--- Search Online English Dictionaries ---]]
> > ... etc
> > You get the idea?

> @pmario No. ... You don't tell me, what exactly should be "removed".

My post should have been clearer :-) What I meant was that **$:/tt/bun/search-en-dictionaries-REMOVER**, in the example might be some kind of self-contained macro that presented the contents of the Bundle List for selective deletion. 
For example ...

```
Select Tiddlers To Delete ...

  [x] $:/tt/bun/search-en-dictionaries
  [ ] $:/tt/bun/search-en-dictionaries-REMOVER
  [ ] $:/tt/eng/en-dict-search.dat
  [ ] $:/tt/eng/en-dict-search.mac
  [x] $:/tt/eng/en-dict-search.hlp
  [ ] $:/tt/eng/en-dict-search.tpl

    [Delete Selected]
```

What I am really looking for is a way to create such a self-contained macro. I'm NOT expecting that is necessarily part of the Bundler plugin. Merely trying to solve a practical issue. 

I hope this is clearer?

P.S. Obviously with Filter Bundles the issue is already solved because of the way they use filters that you access in Advanced Search and can delete via it if needed. The issue is only List Bundles where you have an arbitrary (random) list of Tiddlers that a filter can't easily match.

**Comment by pmario** [comment]
>Select Tiddlers To Delete ...
>
>  [x] $:/tt/bun/search-en-dictionaries
  [ ] $:/tt/bun/search-en-dictionaries-REMOVER
  [ ] $:/tt/eng/en-dict-search.dat
  [ ] $:/tt/eng/en-dict-search.mac
  [x] $:/tt/eng/en-dict-search.hlp
  [ ] $:/tt/eng/en-dict-search.tpl
>
>    [Delete Selected]

I could think about a "remover UI" in the tiddler "info-area", which would allow to do this for advanced users. 

There are 3 possibilities ... 

 1 delete the tiddler but keep it in the bundle
 2 delete the tiddler and remove it from the bundle
 3 keep the tiddler but remove it from the bundle

I'm not sure, what you intended?

**Reply by pmario** [reply]
With a config option the "remover UI" could be "moved" to the bottom of the tiddler, depending on the tiddler tag. 

"Standard bundles" would need a different internal handling than "filtered bundles"

**Reply by TiddlyTweeter** [reply]
> @pmario I could think about a "remover UI" in the tiddler "info-area", which would allow to do this for advanced users.
> 
> There are 3 possibilities ...
> 
> 1 delete the tiddler but keep it in the bundle
> 2 delete the tiddler and remove it from the bundle
> 3 keep the tiddler but remove it from the bundle
> 
> I'm not sure, what you intended?

Thanks for taking my request seriously  and thinking about it!

For MY practical usage I would want to ...
> **2 delete the tiddler and remove it from the bundle**

**Why 2?** _Because any bundle left with contents listed would work for Export still for items that still exist._
I would not want to have either "non-existent" tiddlers listed (1) or bundles that don't list items (3). 
Anyway in case (3) you can easily do it by manual editing a bundle if that is needed.   

That is just MY use case. Others may differ?

**Reply by TiddlyTweeter** [reply]
> @pmario ... With a config option the "remover UI" could be "moved" to the bottom of the tiddler, depending on the tiddler tag.
> 
> "Standard bundles" would need a different internal handling than "filtered bundles"

That sounds excellent.

FYI, in my own use cases I don't need it for Filtered Bundles really as a delete mechanism is already there via Advanced Search. 
It is really only List Bundles where I have real practical need.

**Reply by pmario** [reply]
>FYI, in my own use cases I don't need it for Filtered Bundles really as a delete mechanism is already there via Advanced Search.

That's right. So filtered bundles info area may include a link to the Advanced Search and "prefill" the search input.

**Comment by pmario** [comment]
>For MY practical usage I would want to ...
>
>>    2 delete the tiddler and remove it from the bundle

Hmmm, That's tricky. TW `enlist` filters can't deal with newLines / whitespace atm. ... 

BUT I'd be interested to implement it. 

I'll need an `xenlist` for my story handling plugin anyway. (which is completely broken atm) ... May be we are able to keep the newLines

**Reply by TiddlyTweeter** [reply]
@TiddlyTweeter 
> > For MY practical usage I would want to ...
> > > 2 delete the tiddler and remove it from the bundle

> @pmario  Hmmm, That's tricky. TW `enlist` filters can't deal with newLines / whitespace atm. ...
> 
> BUT I'd be interested to implement it.
> 
> I'll need an `xenlist` for my story handling plugin anyway. (which is completely broken atm) ... May be we are able to keep the newLines

Do you mean that with the current `enlist` filter Tiddler links in a bundle list can be removed/deleted but that the terminating **"\n"** at line end in the bundle list can't be removed? IF so, I don't mind if the number of **blank** lines in a List Bundle increases!

**Reply by pmario** [reply]
We can not create empty lines, which decreases readabiliy

**Reply by TiddlyTweeter** [reply]
> We can not create empty lines, which decreases readabiliy

OK. _I defer to your skill & your want to get it right!_ I was merely commenting I could live with blank lines if there is no other way --- I try not to put up too many obstacles :-).

**Reply by pmario** [reply]
That's the problem. There will be **no** blank lines atm.

**Comment by pmario** [comment]
I'll upload a very basic "info - bundle - tab", that is _only_ able to delete tiddlers. It doesn't remove them from the bundle yet.

**Comment by pmario** [comment]
@TiddlyTweeter see: ~~https://wikilabs.github.io/editions/bundler/experiment-01-26-1652.html~~ .. 

Functions

  - delete the tiddler but keep it in the bundle
  - It contains the comment prefix, and 
 - a new tab in the tiddler info area. 

There is a tiddler : ~~https://wikilabs.github.io/editions/bundler/experiment-01-26-1652.html#future-comment.bundle~~ which contains a different approach for comments. .. I think it will be easier to implement. BUT I don't know yet.

**Reply by pmario** [reply]
WARNING - Do not use this for production. It overwrites quite a bit of the core modules. !!!!

- You should also test the new import mechanism!!!

**Reply by TiddlyTweeter** [reply]
### Overall: Excellent!

> @pmario see: ~~https://wikilabs.github.io/editions/bundler/experiment-01-26-1652.html~~ ..

Thanks! **It works well.** On testing I downloaded your wiki and IMPORTED a bundle of 40 I use regularly into it. Import worked fine and honoured my comments (using method **[[--- comment ---]]**).

**I have changed my mind!!!** on need to delete the item from the Bundle list. It is GOOD in the approach you have taken! Delete from wiki, leave on list. **It works for my use cases well**! Also, IF a user just wants to edit in the bundle itself they never need to look at the info panel.

**There is also an UPSIDE that is not obvious at first**. For instance, say I have created a bundle---but the count does not add up (e.g. **39/40**). Using the INFO panel lets me see WHICH item is missing! It is very helpful to be able to do that!

**Reply by TiddlyTweeter** [reply]
### Regarding Documentation Of TW Settings ...

> @pmario see: ~~https://wikilabs.github.io/editions/bundler/experiment-01-26-1652.html~~ ..

One thing that makes quite a lot of _difference to practical use_ of the info panel is TW settings! **These may be worth mentioning in docs** for the Bundler? In particular I made the following settings active to aid use of the bundler. Eventually I will make a macro to toggle those settings on and off.

> 1 - Set $:/config/MissingLinks to "no"
> 
> 2 - Set $:/config/TiddlerInfo/Default to "$:/core/ui/TiddlerInfo/Bundle".
> 
> 3 - Set $:/config/TiddlerInfo/Mode to "sticky"

**Reply by TiddlyTweeter** [reply]
### Regarding Commenting ...

> @pmario There is a tiddler : https://wikilabs.github.io/editions/bundler/experiment-01-26-1652.html#future-comment.bundle which contains a different approach for comments. .. I think it will be easier to implement. BUT I don't know yet.

I like the idea of leading double-slash   **//**  being for comments, if it can be made to work. Why? Ultimately it is preferable to fake **[[--- Pseudo Tiddlers ---]]**

That said, all my bundles currently use the **[[--- Pseudo Tiddlers ---]]** method, and it works fine.

IF you manage to get **//** working I WOULD use it and phase out **[[--- Pseudo Tiddlers ---]]**.

From my point of view, in the back of my mind, is large numbers of users use the Bundler and a SHARED CONVENTION on COMMENTS I think would be an added plus factor!

**Reply by TiddlyTweeter** [reply]
### A Footnote On Line Spacing
> @pmario see: https://wikilabs.github.io/editions/bundler/experiment-01-26-1652.html ..

Example Bundle ...
```
[[--- This bundle! ---]]
$:/tt/bun/base-1-bundler

[[--- BUNDLER ---------------------------------]]
[[--- install BEFORE other add-ons! ---]]
$:/plugins/wikilabs/bundler
$:/config/wikilabs/enableBundles
$:/config/wikilabs/enableImportBundle
[[--- base Filter Bundles ---]]
$:/tt/bun/all-non-system

[[--- SYSTEM SETTINGS --------------------------]]
[[--- set to "$:/core/ui/TiddlerInfo/Bundle" ---]]
$:/config/TiddlerInfo/Default
[[--- set to "no" ---]]
$:/config/MissingLinks
[[--- set to "sticky" ---]]
$:/config/TiddlerInfo/Mode
```
Under Info renders as ...
![Screenshot 2021-01-27 125758](https://user-images.githubusercontent.com/28830541/105988097-7988b100-609f-11eb-85ae-5766314aa112.jpg)

The info panel works fine WITHOUT BLANK lines, as I hope is clear (though setting $:/config/MissingLinks = "no" helps). It is only in the plain text Bundle List itself that I need them.

Blank lines is not such a major issue really???

**Reply by TiddlyTweeter** [reply]
@pmario The ONLY issue in the example I gave is that COMMENTS have a TRASH BIN button next to them. Since they NEVER exist it could be slightly confusing to a new user?? But I consider that a COSMETIC issue, not a major problem.

Ok. I'm done :-)

**Comment by pmario** [comment]
Using the `[[--- comment ---]]` tiddler convention works without the need to modify or create a new filter-operator. We can use the default "enlist" operator, which IMO is a huge advantage. 

**So I want to keep the convention** ... for now. 

---------

>The ONLY issue in the example I gave is that COMMENTS have a TRASH BIN button next to them. Since they NEVER exist it could be slightly confusing to a new user?? But I consider that a COSMETIC issue, not a major problem.

Yes, that's a problem. I think I can fix that

**Reply by TiddlyTweeter** [reply]
> Using the `[[--- comment ---]]` tiddler convention works without the need to modify or create a new filter-operator. We can use the default "enlist" operator, which IMO is a huge advantage.
> 
> **So I want to keep the convention** ... for now.

**Understood!** Good for me as that is what I've been using for over 2 years already :-)

**Comment by pmario** [comment]
Removing the "trash bin" also looks good with "missing links" enabled now

![grafik](https://user-images.githubusercontent.com/374655/106049604-9eecdd80-60e6-11eb-9cdf-afa10df36026.png)

**Reply by TiddlyTweeter** [reply]
> @pmario Removing the "trash bin" also looks good with "missing links" enabled now

**It does!** Looks like it solves the problem well. 
I'm looking forward to using that once it is available! :-). 

How long should I wait before using the **new** Bundler in production wiki? I  want to update all my wikis with it.

**Reply by pmario** [reply]
I'll release a new version soon today

**Comment by pmario** [comment]
@TiddlyTweeter Released: V0.2.0 of https://wikilabs.github.io/editions/bundler/

**Reply by pmario** [reply]
Video is in the making!

**Reply by TiddlyTweeter** [reply]
@pmario Many thanks! Very appreciated by me!

I hope it will both increase and improve use of The Bundler. 
It is my **No. 1 plugin** (il prima). (IF I were making editions I'd always include it).

**Reply by TiddlyTweeter** [reply]
@pmario 
> ... Released: V0.2.0 of https://wikilabs.github.io/editions/bundler/

I been using it all day. Very neat. **Including Filter Bundles** in the mix was a great idea too. Having selective delete in their info panel is a real plus!  
Basically one can now manage bundles very efficiently. Great tool!

In  English we say "_I am a happy camper!_"

**Reply by TiddlyTweeter** [reply]
> @TiddlyTweeter Released: V0.2.0 of https://wikilabs.github.io/editions/bundler/

In practice you may be interested that the "**import.bundle**", because it too supports deletion, becomes incredibly useful in managing items you no longer need, OR are missing and you need to know what they are.

**Comment by TiddlyTweeter** [comment]
@pmario I have a question about Discussion threads on GitHub! 

Often I have to click on "**View more**"  to actually reveal _many parts_ of the full discussion.
This is frustrating as _often_ what I need to read is inside the reveal!

QUESTION: _Is there a way to expand the entire discussion thread?_

**Reply by pmario** [reply]
You can +1 this one:  "It's frustrating, to click a "view more" button, to reveal valuable conent #3286 " https://github.com/github/feedback/discussions/3286

**Reply by TiddlyTweeter** [reply]
Did so.

**Comment by TiddlyTweeter** [comment]
Just FYI.

I used THIS thread as a very good example of using GitHub Discussions on the [TW Google Group](https://groups.google.com/g/tiddlywiki/c/KldyBhnEYSQ/m/gxDc-layBAAJ).

**Reply by pmario** [reply]
I've seen it :)

**Comment by TiddlyTweeter** [comment]
### I used the Bundler well to show issues with Plugin Libraries ...

[... here ...](https://github.com/Jermolene/TiddlyWiki5/discussions/5524#discussion-3250504 )
